### ISSUE TITLE
question: is this dbus signal handling code idiomatic?

### ISSUE THREAD

**Issue by stapelberg** [diwic/dbus-rs#46]
I’m prototyping a program which communicates with Avahi via dbus. Avahi sends signals for newly discovered services, so my code looks like this:

```rust
extern crate dbus;

#[derive(Debug)]
struct ServiceBrowserItemNew {
    interface: i32,
    protocol: i32,
    name: String,
    item_type: String,
    domain: String,
    flags: u32,
}

fn service_browser_item_new_msg(m: &dbus::Message) -> Option<ServiceBrowserItemNew> {
    let mut iter = m.iter_init();
    let interface = iter.get::<i32>().unwrap();

    assert!(iter.next());
    let protocol = iter.get::<i32>().unwrap();

    assert!(iter.next());
    let name = iter.get::<&str>().unwrap().to_string();

    assert!(iter.next());
    let item_type = iter.get::<&str>().unwrap().to_string();

    assert!(iter.next());
    let domain = iter.get::<&str>().unwrap().to_string();

    assert!(iter.next());
    let flags = iter.get::<u32>().unwrap();

    Some(ServiceBrowserItemNew {
        interface: interface,
        protocol: protocol,
        name: name,
        item_type: item_type,
        domain: domain,
        flags: flags,
    })
}
fn focus_msg(ci: &dbus::ConnectionItem) -> Option<ServiceBrowserItemNew> {
    let m = if let &dbus::ConnectionItem::Signal(ref s) = ci {
        s
    } else {
        return None;
    };

    if &*m.interface().unwrap() == "org.freedesktop.Avahi.ServiceBrowser" &&
       &*m.member().unwrap() == "ItemNew" {
        return service_browser_item_new_msg(m);
    }

    None
}

fn main() {
    let c = dbus::Connection::get_private(dbus::BusType::System).unwrap();
    let sbn = dbus::Message::new_method_call("org.freedesktop.Avahi",
                                             "/",
                                             "org.freedesktop.Avahi.Server",
                                             "ServiceBrowserNew").unwrap()
    .append1(-1i32) // interface: AVAHI_IF_UNSPEC
    .append1(-1i32) // protocol: AVAHI_PROTO_UNSPEC (ipv4+ipv6)
    .append1("_googlecast._tcp") // type
    .append1("") // TODO: is "" == null?
    .append1(0u32); // flags: 0

    let r = c.send_with_reply_and_block(sbn, 2000).unwrap();
    let path: dbus::Path = r.get1().unwrap();
    println!("returned path = {}", path);

    c.add_match("interface='org.freedesktop.Avahi'").unwrap();

    for i in c.iter(1000) {
        if let Some(app) = focus_msg(&i) {
            // e.g. ServiceBrowserItemNew { interface: 3, protocol: 0,
            // name: "Chromecast", item_type: "_googlecast._tcp", domain: "local", flags: 5 }
            println!("got a message: {:?}", app);
        } else {
            println!("got message which i do not understand");
        }
    }
}
```

The signal decoding part seems pretty cumbersome to me. I’m new to Rust, so it’s pretty likely that I’m missing something, hence I figured I’d ask.

In Go, I can imagine a design using reflection and tags (like JSON and XML decoding work):
```go
type ServiceBrowserItemNew struct {
  Interface int32 `dbus:"1"`
  Protocol  int32 `dbus:"2"`
  // …
}

func main() {
  // …
  for {
    msg := <-conn.Items()
    if msg.Interface == "org.freedesktop.Avahi.ServiceBrowser" && msg.Member == "ItemNew" {
      var sbin ServiceBrowserItemNew
      if err := msg.Unmarshal(&sbin); err != nil {
        log.Fatal(err)
      }
      // …
    }
  }
}
```

Is something similar possible in Rust with this dbus library? Or is iterating through the individual message arguments really the idiomatic way?

**Comment by diwic** [comment]
> Is something similar possible in Rust with this dbus library? 

Not currently, but that's mostly because nobody has implemented it yet. One problem is that Rust is torn between the older "rustcserialize" and the newer "serde", and I have had problems making up my mind w r t which one to support (or both).

 > Or is iterating through the individual message arguments really the idiomatic way?

Yes - that's basically what you do - or use `get1`, `get2`, or `get3`, but you would need a `get6` so it doesn't work for you.

What looks a bit inconsistent in your code is that you return a `Option<ServiceBrowserItemNew>` but there is no way to return `None` from the function - instead you panic on message not matching.
Whether to unwrap or to do proper error handling is up to you, but this seems like a mix.

You do raise a valid point in that this could be done in a more ergonomic way. I'll give this a thought.

**Comment by diwic** [comment]
Ok, so I've added an `iter::read` method now. Using it your code would look like:

    fn service_browser_item_new_msg(m: &dbus::Message) -> Result<ServiceBrowserItemNew, TypeMismatchError> {
       let mut iter = m.iter_init();
       Ok(ServiceBrowserItemNew {
           interface: try!(iter.read()),
           protocol: try!(iter.read()),
           name: try!(iter.read::<&str>()).into(),
           item_type: try!(iter.read::<&str>()).into(),
           domain: try!(iter.read::<&str>()).into(),
           flags: try!(iter.read()),
       })
    }

Could you give it a try and see if it works the way I expect it to? If so I might do a new release with this function in (and might even steal your struct definition and use as an example, with your permission).

**Comment by stapelberg** [comment]
Thanks a bunch for the quick reply and fix! I’ll be happy to test that tomorrow. Do you happen to have a pointer as to how I can use the git version of your dbus library in my Cargo.toml in a way that includes the change you mentioned?

**Comment by diwic** [comment]
Like this:

    [dependencies]
    dbus = { git = "https://github.com/diwic/dbus-rs" }

**Comment by stapelberg** [comment]
Thanks a bunch! I can confirm that the following works for me:

```rust
fn service_browser_item_new_msg(m: &dbus::Message)
                                -> Result<ServiceBrowserItemNew, dbus::arg::TypeMismatchError> {
    let mut iter = m.iter_init();
    Ok(ServiceBrowserItemNew {
        interface: try!(iter.read()),
        protocol: try!(iter.read()),
        name: try!(iter.read()),
        item_type: try!(iter.read()),
        domain: try!(iter.read()),
        flags: try!(iter.read()),
    })
}
```

Feel free to use that in the documentation.

**Comment by diwic** [comment]
:+1:  

Now released as v0.3.4.
